enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,825 @@
+* Eventually[^3] remove the Machine API controllers and leverage the migration layer to convert Machine API machines to Cluster API machines
+* Allow the Cluster Autoscaler, MachineHealthCheck and ControlPlaneMachineSet to continue to work during a semi-migrated cluster state
+* Ensure that every Machine API resource is translated to a Cluster API resource to allow users to test the new API
+* Allow access to newer features of Cluster API that do not exist in Machine API
+* Avoid carry patches in the Cluster API controllers to ensure that we can easily upgrade to newer versions of Cluster API
+
+[^3]: In the first iteration of this project, the goal is to allow users to migrate individual resources.
+In later iterations, the plan is to remove the Machine API controllers. The linked goals will be executed in separate releases.
+
+### Non-Goals
+
+* Setting up or configuring any part of Cluster API, this is handled in a separate enhancement
+* Modifying any of the existing deployments of Machine API within clusters
+* Deleting or modifying any existing resources within clusters (until a user chooses to do so)
+* Force creation of new Machine API machines when new Cluster API MachineSets are being created[^4]
+* Port any feature from Cluster API into Machine API
+
+[^4]: We want to promote, over time, a reduction in the number of Machine API Machines, as such,
+any resources created directly in Cluster API, will not be reflected into Machine API, unless they have an owner,
+and, that owner is already reflected into Machine API.
+
+## Proposal
+
+We will implement a two-way sync controller that synchronises Machines and related resources between the Machine API and Cluster API equivalents.
+
+Using a new field on the Machine API resource, users will be able to choose which API is authoritative.
+Any discrepancy between the non-authoritative resource and the authoritive resource will be overwritten by the sync controller to match the authoritative resource.
+
+When the API is non-authoritative, the controllers should be paused (ignore the resource),
+allowing the authoritative API’s controller to perform the reconciliation actions required.
+
+### Workflow Description
+
+In this workflow, the **cluster admin** is responsible for the infrastructure provisioning for the cluster.
+
+When the cluster admin wishes to migrate a MachineSet from Machine API to Cluster API, the following procedure is required:
+1. Identify the MachineSet to migrate to Cluster API
+1. Use the `oc edit` or a patch command to update the value of the `spec.authoritativeAPI` field to `cluster.x-k8s.io`
+1. The sync controller verifies that the move from Machine API to Cluster API is valid and that the Cluster API version of the resource is fully synced with the Machine API version
+1. The `status.authoritativeAPI` field is updated to `transitioning.to.cluster.x-k8s.io` and the `status.synchronizedGeneration` is updated to the current generation of the authoritative resource
```

Comment:
Not sure I like `transitioning.to.cluster.x-k8s.io`, perhaps better to use `Migrating` and rely on the spec for the direction 🤔